pkg/packet/bgp: use netip for IPAddrPrefix and IPv6AddrPrefix

fujita opened the pull request:
Replace uses of net.IP with netip.Prefix for IPAddrPrefix and IPv6AddrPrefix.

When GoBGP was started, the net/netip package did not exist and net.IP was the standard type for representing IP addresses. However, with the introduction of netip.Addr in Go 1.18, we now have a more efficient and type-safe alternative.

Although netip.Addr may use slightly more memory for IPv4 addresses (as it always reserves space for IPv6), it eliminates the need for memory allocation on the heap, which is expected to improve performance, especially in high-throughput paths.

This patch replaces uses of net.IP with netip.Prefix type implemented by using netip.Addr for IPAddrPrefix and IPv6AddrPrefix.

We should use netip for IpAddrPrefixDefault however it means that this patch becomes much larger since we also update all labeled addrprefix types. There will be some temporary code duplication, but I plan to gradually proceed with the replacement.

maxime-peim commented:
> Although netip.Addr may use slightly more memory for IPv4 addresses (as it always reserves space for IPv6), it eliminates the need for memory allocation on the heap, which is expected to improve performance, especially in high-throughput paths.

Compared to the [net.IP](https://pkg.go.dev/net#IP) type, [Addr](https://pkg.go.dev/net/netip#Addr) type takes less memory, is immutable, and is comparable (supports == and being a map key). [1]

[1]https://pkg.go.dev/net/netip

Also, `net.IP` also allocated memory for IPv6 when building an IPv4: https://cs.opensource.google/go/go/+/refs/tags/go1.24.5:src/net/ip.go;l=53

> it eliminates the need for memory allocation on the heap

That is totally false. When the variable needs to outlive the scope of the function it is declared in, it is always declared on the heap.

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1565,8 +1566,36 @@ func (r *IPAddrPrefix) DecodeFromBytes(data []byte, options ...*MarshallingOptio
+		return NewMessageError(eCode, eSubCode, nil, msg)
+	}
+
+	bytelen := (int(bitlen) + 7) / 8
+	if len(data) < bytelen {
+		return nlriError("network bytes is short")
+	}
+	if bitlen > r.addrlen*8 {
+		return nlriError("network bit length is too long")
+	}
+
+	var b [16]byte
+	copy(b[:], data[:bytelen])
+	// clear trailing bits in the last byte. rfc doesn't require
+	// this but some bgp implementations need this...
+	rem := bitlen % 8
+	if rem != 0 {
+		mask := 0xff00 >> rem
+		lastByte := b[bytelen-1] & byte(mask)
+		b[bytelen-1] = lastByte
+	}
+	addr, _ := netip.AddrFromSlice(b[:r.addrlen])
+	r.Prefix = netip.PrefixFrom(addr, int(bitlen))
+	return nil
```
It should be moved in a separate function for future use with `netip`

fujita replied:
You mean it should be a separate function and it will be used other IPAddrPrefix structures like LabeledVPNIPAddrPrefix?
If so, as I wrote in the commit message, this structure will be IPAddrPrefixDefault, will be embedded in other structures. So no need to make this a separate function.

maxime-peim replied:
It is a bit weird to update an other type to become the new default, would it be clearer to have a new type `IPAddrPrefixDefaultNetip` to be used incrementally in other structures?

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1582,11 +1611,13 @@ func (r *IPAddrPrefix) Serialize(options ...*MarshallingOption) ([]byte, error)
-	if err != nil {
-		return nil, err
-	}
+	bits := uint8(r.Prefix.Bits())
+	buf = append(buf, bits)
+
+	byteLen := (int(bits) + 7) / 8
+	pbuf := make([]byte, byteLen)
+	copy(pbuf, r.Prefix.Addr().AsSlice()[:byteLen])
```
It should be moved in a separate function for future use with netip

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1599,19 +1630,48 @@ func (r *IPAddrPrefix) SAFI() uint8 {
-			Length: length,
-		},
-		4,
+	if prefix != "" {
+		addr := netip.MustParseAddr(prefix)
+		addrlen := net.IPv4len
+		if addr.Is6() {
+			addrlen = net.IPv6len
+		}
+
+		p := &IPAddrPrefix{
+			addrlen: uint8(addrlen),
+		}
+		buf := append([]byte{length}, addr.AsSlice()[:addrlen]...)
+		err := p.DecodeFromBytes(buf)
+		if err != nil {
+			return nil
+		}
+		return p
+	} else {
+		return &IPAddrPrefix{}
 	}
```
```go
if prefix == "" {
	return &IPAddrPrefix{}
}

addr := netip.MustParseAddr(prefix)
addrlen := net.IPv4len
if addr.Is6() {
	addrlen = net.IPv6len
}

p := &IPAddrPrefix{
	addrlen: uint8(addrlen),
}
buf := append([]byte{length}, addr.AsSlice()[:addrlen]...)
err := p.DecodeFromBytes(buf)
if err != nil {
	return nil
}
return p
```

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1599,19 +1630,48 @@ func (r *IPAddrPrefix) SAFI() uint8 {
-		},
-		4,
+	if prefix != "" {
+		addr := netip.MustParseAddr(prefix)
```
you shouldn't use the `Must` version to avoid panic-ing

fujita replied:
NewSomePrefix APIs don't return an error. So for me, you must pass a valid string to those APIs.

maxime-peim replied:
I am not comfortable with that. If a user mistype a prefix from the CLI for example, I don't think it is a good idea to make the whole gobgp process panic. You should handle and report an error.

fujita replied:
If you prefer, I'm fine with returning error. It should be done in a different commit because we change the API interface.

maxime-peim replied:
```go
func NewIPAddrPrefix(length uint8, prefix string) *IPAddrPrefix {
	p := &IPAddrPrefix{
		IPAddrPrefixDefault{
			Length: length,
		},
		4,
	}
	// TODO: pass the error to the caller
	_ = p.decodePrefix(net.ParseIP(prefix).To4(), length, 4)
	return p
}
```
if `prefix` is not valid, you get a `nil` value from `ParseIP`, `To4` returns `nil` as well and an error is returned.
As I wrote in a previous PR, `TODO: pass the error to the caller`, we should handle that error to avoid panic. I don't think that explicitly panic is the right way to do it, you should pass the error to the caller where it should be handled or passed upper.

fujita replied:
Ok, added this to the TODO list
https://github.com/osrg/gobgp/issues/3043

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1599,19 +1630,48 @@ func (r *IPAddrPrefix) SAFI() uint8 {
+		p := &IPAddrPrefix{
+			addrlen: uint8(addrlen),
+		}
+		buf := append([]byte{length}, addr.AsSlice()[:addrlen]...)
+		err := p.DecodeFromBytes(buf)
+		if err != nil {
+			return nil
+		}
```
one the prefix decoding is moved to its own function, you should directly use it as it was done before

fujita replied:
As I wrote above, I plan to make `NewIPAddrPrefix()` take only a valid string. So this function will not need DecodeFromBytes stuff. It just calls `netip.MustParseAddr` and then `PrefixFrom`.

Note that I plan to change the `DecodeFromBytes` API for AddrPrefix.
As a premise, I believe that the same struct should be used for both v6 and v4 like netip, and I'm planning to make that change.

Currently, we create an uninitialized `IPAddrPrefix`, set the addrlen for v6 cases, and then call `DecodeFromBytes`. Instead, I like to have something like `func IPAddrPrefixFromBytes(family, []byte) (IPAddrPrefix, error)`.  The API interface will be aligned with netip's API. Unfortunately, Go doesn't enforce RAII though.

`netip.Addr` has the information about ipv4 or v6 so we don't need addrlen stuff.

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1599,19 +1630,48 @@ func (r *IPAddrPrefix) SAFI() uint8 {
+		addr := netip.MustParseAddr(prefix)
+		addrlen := net.IPv4len
+		if addr.Is6() {
+			addrlen = net.IPv6len
```
this is a big change, you allow ipv6 addresses to be decoded from there, which was not the case before

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1626,26 +1686,14 @@ func (r *IPv6AddrPrefix) AFI() uint16 {
-				Length: length,
-			},
-			16,
+	p := *NewIPAddrPrefix(length, prefix)
```
you don't need to dereference the return value, but you should check that it is not nil.

maxime-peim replied:
as stated before, panic-ing is not a good solution, errors happens and should be reported and not stop the whole process

fujita replied:
As I wrote above, I believe that the current API panics with an invalid input. So I don't change the semantics.

maxime-peim replied:
That is not true, the current does not panic directly, you assumed it will panic at some point but it will not in the `New*Prefix` function

fujita replied:
> That is not true, the current does not panic directly, you assumed it will panic at some point but it will not in the `New*Prefix` function

Ok, I keep that semantics, it will panic at some point. From user point of view, I think that panicking immediately is better though.

fujita commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1565,8 +1566,36 @@ func (r *IPAddrPrefix) DecodeFromBytes(data []byte, options ...*MarshallingOptio
+	bitlen := data[0]
+	data = data[1:]
+
+	nlriError := func(msg string) error {
+		eCode := uint8(BGP_ERROR_UPDATE_MESSAGE_ERROR)
+		eSubCode := uint8(BGP_ERROR_SUB_INVALID_NETWORK_FIELD)
+		return NewMessageError(eCode, eSubCode, nil, msg)
+	}
```
I thought that it was more readable but yeah probably not. I'll change.

fujita commented:
> > Although netip.Addr may use slightly more memory for IPv4 addresses (as it always reserves space for IPv6)
> 
> From https://pkg.go.dev/net/netip: Compared to the [net.IP](https://pkg.go.dev/net#IP) type, [Addr](https://pkg.go.dev/net/netip#Addr) type takes less memory, is immutable, and is comparable (supports == and being a map key).
> 
> Also, `net.IP` also allocates memory for IPv6 when building an IPv4: https://cs.opensource.google/go/go/+/refs/tags/go1.24.5:src/net/ip.go;l=53

Ah, good. Then there are no disadvantages to migrating from net to netip.
 
> > it eliminates the need for memory allocation on the heap
> 
> That is totally false. When the variable needs to outlive the scope of the function it is declared in, it is always declared on the heap.

Hmm, I was talking about the number of heap allocations.

```
type A struct {
   len uint32
   add netip.Addr
}

type B struct {
   len uint32
   addr net.IP
}
```
If we have something like above, there could be two heap allocations for struct B. On the other hand, struct A do one heap allocation. As you said, if net.IP always allocates enough space for v6, we could do one heap allocation for struct B too though.

fujita commented:
@maxime-peim As you suggested, I go with the approach to add `IPAddrPrefixDefaultNetip` temporarily. As a result, I should have addressed all of your review comments.

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1540,15 +1540,61 @@ func (r *IPAddrPrefixDefault) MarshalJSON() ([]byte, error) {
-type IPAddrPrefix struct {
-	IPAddrPrefixDefault
+// Once we convert all the users of IPAddrPrefixDefault to use IPAddrPrefixDefaultNetip,
+// this is renamed.
```
`this will be renamed`

fujita replied:
Oops, will fix.

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1540,15 +1540,61 @@ func (r *IPAddrPrefixDefault) MarshalJSON() ([]byte, error) {
+		return NewMessageError(BGP_ERROR_UPDATE_MESSAGE_ERROR, BGP_ERROR_SUB_INVALID_NETWORK_FIELD, nil, "network bit length is too long")
+	}
+
+	var b [16]byte
```
why do you differ from `IPAddrPrefixDefault.decodePrefix` there?
since it is exactly the same code, and just the two last lines differ when setting the `Prefix` field to the right type, can't you create an underlying function that takes `data []byte, bitlen uint8, addrlen uint8` and returns the parsed buffer?
this would avoid a lot of code duplication

maxime-peim replied:
```go
func decodePrefix(data []byte, bitlen uint8, addrlen uint8) ([]byte, error) {
	bytelen := (int(bitlen) + 7) / 8
	if len(data) < bytelen {
		eCode := uint8(BGP_ERROR_UPDATE_MESSAGE_ERROR)
		eSubCode := uint8(BGP_ERROR_SUB_MALFORMED_ATTRIBUTE_LIST)
		return nil, NewMessageError(eCode, eSubCode, nil, "network bytes is short")
	}
	if bitlen > addrlen*8 {
		eCode := uint8(BGP_ERROR_UPDATE_MESSAGE_ERROR)
		eSubCode := uint8(BGP_ERROR_SUB_MALFORMED_ATTRIBUTE_LIST)
		return nil, NewMessageError(eCode, eSubCode, nil, "network bit length is too long")
	}
	b := make([]byte, addrlen)
	copy(b, data[:bytelen])
	// clear trailing bits in the last byte. rfc doesn't require
	// this but some bgp implementations need this...
	rem := bitlen % 8
	if rem != 0 {
		mask := 0xff00 >> rem
		lastByte := b[bytelen-1] & byte(mask)
		b[bytelen-1] = lastByte
	}
	return b, nil
}

func (r *IPAddrPrefixDefault) decodePrefix(data []byte, bitlen, addrlen uint8) error {
	b, err := decodePrefix(data, bitlen, addrlen)
	r.Prefix = b
	return err
}

func (r *IPAddrPrefixDefaultNetip) decodePrefix(data []byte, bitlen, addrlen uint8) error {
	b, err := decodePrefix(data, bitlen, addrlen)
	if err != nil {
		return err
	}
	addr, _ := netip.AddrFromSlice(b[:r.addrlen])
	r.Prefix = netip.PrefixFrom(addr, int(bitlen))
	return nil
}
```

The behavior of `IPAddrPrefixDefault` does not change and you avoid copying the same code

maxime-peim replied:
This cannot breaks the usage of `IPAddrPrefixDefault` since you do exactly the same thing.

maxime-peim commented on pkg/packet/bgp/bgp.go:
```diff
@@ -1540,15 +1540,61 @@ func (r *IPAddrPrefixDefault) MarshalJSON() ([]byte, error) {
 
-func (r *IPAddrPrefix) DecodeFromBytes(data []byte, options ...*MarshallingOption) error {
+func (r *IPAddrPrefixDefaultNetip) decodePrefix(data []byte, bitlen uint8) error {
 	if r.addrlen == 0 {
```
it is weird to check `addrlen` here and in `DecodeFromBytes`, both of the check should be done in `DecodeFromBytes`

fujita replied:
Why weird? IPAddrPrefixDefaultNetip owns addrlen so checking it inside the method is ok for me. Once we merge v6 and v4 structures, it will be removed. So it doesn't matter much.

maxime-peim replied:
```go
func (r *IPAddrPrefix) DecodeFromBytes(data []byte, options ...*MarshallingOption) error {
	f := RF_IPv4_UC
	switch r.addrlen {
		case 0:
			r.addrlen = net.IPv4len
		case net.IPv6len:
			f = RF_IPv6_UC
		default:
			return ...
	}

	if IsAddPathEnabled(true, f, options) {
		var err error
		data, err = r.decodePathIdentifier(data)
		if err != nil {
			return err
		}
	}
	if len(data) < 1 {
		eCode := uint8(BGP_ERROR_UPDATE_MESSAGE_ERROR)
		eSubCode := uint8(BGP_ERROR_SUB_MALFORMED_ATTRIBUTE_LIST)
		return NewMessageError(eCode, eSubCode, nil, "prefix misses length field")
	}
	r.Length = data[0]
	return r.decodePrefix(data[1:], r.Length, r.addrlen)
}
```

So you don't have to check and modify the addrlen value in decode prefix, and you keep the same signature for `decodePrefix` as it was before.

maxime-peim replied:
See: 
```go
if r.addrlen == 0 {
	r.addrlen = 4
```
move from `DecodeFromBytes` to `decodePrefix` for no reason
